### PR TITLE
Master

### PR DESCRIPTION
**Pull request by porkybrain**

### CODE REVIEW

**Comment by porkybrain** [comment]
I added a bunch of missing tags and inheritance for inheriting peripherals and register defaults. This allows for adiquate use of the nordic chips which actually heavily use inheritance.

I implemented the default value propagation as a free function taking an SVDDevice. I assume this will raise a discussion about ergonomics again. 

As far as I understand the `SVDParser` class is esentailly a factory pattern which also happends to store an xml tree. Essentailly I could use any property tree if I just use the naked constructor. As of now you can decide if you want fields named reserved to be removed automatically and if you want to expand arrays of registers or not. Ignoring "reserved" is somethinge which we should have just implemented to begin with since it says here http://www.keil.com/pack/doc/CMSIS/SVD/html/svd__outline_pg.html (see field level) to do it. I think this should at least be on by default (although this may be a breaking change to leagacy users).

The arguement was raised that with register duplucation we should allow the user to choose to do it either at parse time or later, hence the free function. With default value propagation simmilar arguments can be raised. So if we are going to implement everything as a flag in our factory and as a free function it seems better to me to just use a free function and be done with it. 

I don't really see the use in `SVDParser` having state at all. Its not like your going to construct it with a tree and then call get_device, loose your device object and need to call get_device again and again. Even if you do then its very inefficient because the device is parsed again and again. 

I'm not an expert on python encapsulation conventions but in my mind the public interface for this library should be:
 - a free function which finds a packaged svd file based on a vendor and file name
 - a free function which does initial parsing on that xml file returning a SVDDevice
 - a free function which expands arrays of registers like the one we have now
 - a free function which propagates default values

This would be much simpler and in my mind more idiomatic. I don't see a functionality loss anywhere.

File: python/cmsis_svd/examples/show_k20_peripherals.py [hunk 1]
```diff
@@ -15,6 +15,6 @@
 #
 from cmsis_svd.parser import SVDParser
 
-parser = SVDParser.for_packaged_svd('Freescale', 'MK20D7.xml')
+parser = SVDParser.for_packaged_svd('Freescale', 'MK20D7.swd')
```
Comment (posborne) [inline python/cmsis_svd/examples/show_k20_peripherals.py]: Shouldn't this be `.svd`?

File: python/cmsis_svd/model.py [hunk 2]
```diff
@@ -126,34 +126,48 @@ def __init__(self, name, value):
...
+        self.access = access                    #Defines the default access rights for all registers. 
+        self.protection = protection            #Defines extended access protection for all registers. 
+        self.reset_value = reset_value          #Defines the default value for all registers at RESET. 
+        self.reset_mask = reset_mask            #Identifies which register bits have a defined reset value. 
         
    
 class SVDCpu(SVDElement):
 
-    def __init__(self, name, revision, endian, mpu_present, fpu_present, vtor_present, nvic_prio_bits, vendor_systick_config):
+    def __init__(self, name, revision, endian, mpu_present, fpu_present, fpu_dp, icache_present, dcache_present, itcm_present, dtcm_present, vtor_present, nvic_prio_bits, vendor_systick_config, device_num_interrupts, sau_num_regions, sau_regions_config):
```
Comment (posborne) [inline python/cmsis_svd/model.py]: Break this up with a newline, getting a bit long

File: python/cmsis_svd/parser.py [hunk 3]
```diff
@@ -66,8 +66,8 @@ class SVDParser(object):
     expand_arrays_of_registers = 0
 
     @classmethod
-    def for_xml_file(cls, path):
-        return cls(ET.parse(path))
+    def for_xml_file(cls, path, remove_reserved = 0, expand_arrays_of_registers = 0):
```
Comment (posborne) [inline python/cmsis_svd/parser.py]: Really minor, kwargs should have no space around the `=` according to PEP8.   E.g. `remove_reserved=0`.

File: python/cmsis_svd/parser.py [hunk 4]
```diff
@@ -66,8 +66,8 @@ class SVDParser(object):
...
+        return cls(ET.parse(path),remove_reserved,expand_arrays_of_registers)
```
Comment (posborne) [inline python/cmsis_svd/parser.py]: Another minor PEP8 things, should be spaces after commas here.

File: python/cmsis_svd/parser.py [hunk 5]
```diff
@@ -243,6 +264,54 @@ def duplicate_array_of_registers(input):    #expects a SVDRegister which is an a
...
+
+    if device.peripherals:
+        for peripheral in device.peripherals:
+            if peripheral.derived_from:
+                parent = None;
+                for p in device.peripherals:
+                    if p.name == peripheral.derived_from:
+                        parent = p      #TODO support dot see http://www.keil.com/pack/doc/CMSIS/SVD/html/svd__outline_pg.html
+                if peripheral.description is None:
+                    peripheral.description = parent.description
```
Comment (posborne) [inline python/cmsis_svd/parser.py]: As a demonstration of an alternative to this approach, see https://github.com/posborne/cmsis-svd/commit/b277979ac7c572e5f8ef7d8ecada5a25b3a47acf.  The other alternative is to store description in a field like `_description` on the Peripheral and then define a property like the following:

```python
@property
def description(self):
    return self._description or self.device.description
```

Both of those use a clever but convenient form of short-circuit evaluation to filter out the None case.

File: python/cmsis_svd/examples/show_k20_peripherals.py (hunk 1 above)
Comment (porkybrain) [inline python/cmsis_svd/examples/show_k20_peripherals.py]: It is now :)

File: python/cmsis_svd/parser.py (hunk 5 above)
Comment (porkybrain) [inline python/cmsis_svd/parser.py]: The problem with short circuiting is that both `None` and `0x00000000` are interpreted as false, therefore anywhere where o is a valid value we must still distinguish between 0 and `None`.

**Comment by posborne** [comment]
I had kind of forgotten about this change.  I'll try to re-review and get this to a mergeable state soonish.

**Comment by MightyPork** [comment]
would love to see this merged, `derived_from` is sorely missing. I'm trying to generate asm file with all registers and it's kinda pain with multiple instances

**Comment by posborne** [comment]
I am going to close this PR and open a new one that includes the original contents from @porkybrain but also includes my changes on top of that.

**Pull request closed without merging**